Adding how-to guide for tf private registry for recipes (#1073)
radius-project/docs: docs/content/guides/recipes/Terraform/howto-private-registry/index.md

Reviewer: This appears to be in a new "Terraform" directory. Could you remove this directory and save this for #1077

weight: 500
categories: "How-To"
tags: ["recipes", "terraform"]
---

This how-to guide will describe how to:

- Configure a Radius Environment to be able to pull Terraform Recipe templates from a private git repository.

### Prerequisites

Before you get started, you'll need to make sure you have the following tools and resources:

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
- [Radius Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- A [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#about-personal-access-tokens) for your Git repository.

## Step 1: Define a secret store resource

Begin by configuring a [Radius Secret Store]({{< ref secretstore >}}) with the personal access token or username + password you previously created, which has access to your private git repository.

Create a Bicep file `env.bicep`, import Radius and  define your resource:

{{< rad file="snippets/env.bicep" embed=true marker="//SECRETSTORE" >}}

> > Note the property `pat` is a required property that refers to your personal access token, while `username` can be optional.

Reviewer: ```suggestion
> Note the property `pat` is a required property that refers to your personal access token, while `username` is an optional field you can specify if your git platform requires a username.
```

## Step 2: Define your Radius Recipe configurations

Radius provides a property `recipeConfig` which allows users to setup specific Git module sources for your Terraform Radius Recipes. In your `env.bicep` file define this property in your Radius Environment resource, visit the [Radius Environment schema]({{< ref environment-schema >}}) page for more information.

{{< rad file="snippets/env.bicep" embed=true marker="//ENV" >}}

> Note the keys listed inside of the `pat` property should match your path name to the secret store.

> ## Step 3: Define your Radius Recipe

Reviewer: ```suggestion
## Step 3: Add a Terraform Recipe
```

> Define your Radius Recipe inside your [Radius Environment resource]({{< ref environment-schema >}}) in your `env.bicep` file, keep in mind that your `templatePath` should contain a `git::` prefix as per [Terraform requirements](https://developer.hashicorp.com/terraform/language/modules/sources#generic-git-repository).

Reviewer: ```suggestion
Update your Environment with a Terraform Recipe, pointing to your private git repository. Note that your `templatePath` should contain a `git::` prefix, per the [Terraform module documentation](https://developer.hashicorp.com/terraform/language/modules/sources#generic-git-repository).
```

---
type: docs
title: "How-To: Pull Terraform modules from private git repositories"
linkTitle: "Private git repos"
description: "Learn how to setup your Radius environment to pull Terraform Recipe templates from a private git repository."
weight: 500
categories: "How-To"
tags: ["recipes", "terraform"]
---

This how-to guide will describe how to:

- Configure a Radius Environment to be able to pull Terraform Recipe templates from a private git repository.

### Prerequisites

Before you get started, you'll need to make sure you have the following tools and resources:

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})

> - [Radius Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})

Reviewer 2: ![image](https://github.com/radius-project/docs/assets/100623239/feb074a4-48a7-4121-a307-7aeaaf9b7ff1)

Formatting seems a bit off here

- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- A [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#about-personal-access-tokens) for your Git repository.

## Step 1: Define a secret store resource

Begin by configuring a [Radius Secret Store]({{< ref secretstore >}}) with the personal access token or username + password you previously created, which has access to your private git repository.

Create a Bicep file `env.bicep`, import Radius, and  define your resource:

{{< rad file="snippets/env.bicep" embed=true marker="//SECRETSTORE" >}}

> > Note the property `pat` is a required property that refers to your personal access token, while `username` is an optional field you can specify if your git platform requires a username.

Reviewer 2: ```suggestion
> Note the property `pat` is a required secret key that refers to your personal access token, while `username` is an optional key you can specify if your git platform requires a username.
```

---
type: docs

> title: "How-To: Set up a private registry for your Terraform Radius Recipes"

Reviewer: ```suggestion
title: "How-To: Pull Terraform modules from private git repositories"
```

> title: "How-To: Setup a private registry for your Terraform Radius Recipes"
> linkTitle: "Setup a private registry"

Reviewer 3: ```suggestion
title: "How-To: Set up a private registry for your Terraform Radius Recipes"
linkTitle: "Set up a private registry"
```

> linkTitle: "Set up a private registry"

Reviewer: ```suggestion
linkTitle: "Private git repos"
```

> description: "Learn how to configure your Radius Environment to leverage your private registry as storage for your custom Recipe templates"

Reviewer: ```suggestion
description: "Learn how to setup your Radius environment to pull Terraform Recipe templates from a private git repository."
```

weight: 500
categories: "How-To"
tags: ["recipes", "terraform"]
---

This how-to guide will provide an overview of how to:

> - Configure a Radius Environment that leverages a [Radius secret store]({{< ref secretstore >}}) to register a [private Git registry](https://developer.hashicorp.com/terraform/registry/private) for your Recipe templates.

Reviewer 3: is this supposed to be `private Terraform registry`?

```suggestion
- Configure a Radius Environment that leverages a [Radius secret store]({{< ref secretstore >}}) to register a [private Terraform registry](https://developer.hashicorp.com/terraform/registry/private) for your Recipe templates.
```

> - Configure a Radius Environment that leverages a [Radius secret store]({{< ref secretstore >}}) to register a [private Terraform registry](https://developer.hashicorp.com/terraform/registry/private) for your Recipe templates.

Reviewer: ```suggestion
This how-to guide will describe how to:

- Configure a Radius Environment to be able to pull Terraform Recipe templates from a private git repository.
```

Private Terraform registries are not the same as private git repositories.

### Prerequisites

Before you get started, you'll need to make sure you have the following tools and resources:

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
- [Radius Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

> 

Reviewer: I'd add a step here for creating a PAT. You can link to the docs pages for some popular platforms like Azure DevOps, GitHub, and GitLab.

### Step 1: Define a secret store resource

> Radius provides support for defining Kubernetes Secrets as Bicep resources, through the Radius Secret Store. Users have the option to both leverage an existing Kubernetes Secret or define a new Kubernetes Secret in the resource, see the [Radius Secret Store schema]({{< ref secretstore >}}) for more information.
> 
> 

Reviewer: ```suggestion
Begin by configuring a [Radius Secret Store]({{< ref secretstore >}}) with the personal access token or username + password you previously created, which has access to your private git repository.
```

Create a Bicep file `env.bicep` and define your resource:

> {{< rad file="snippets/env.bicep" embed=true marker="//SECRETSTORE" >}}

Reviewer: This will cause errors because you didn't instruct the user to import radius

> Note the property `pat` is a required property that refers to your personal access token, while `username` can be optional.

> ## Step 2: Define your Radius Recipe

Reviewer: I think this step should be after the Recipe configuration instead of before

> Define your Radius Recipe, keep in mind that your `templatePath` should contain a `git::` prefix as per [Terraform requirements](https://developer.hashicorp.com/terraform/language/modules/sources#generic-git-repository).

Reviewer 3: Do they define the recipe in the `env.bicep` file as well? If so, please specify that as well.

{{< rad file="snippets/env.bicep" embed=true marker="//RECIPE" >}}

## Step 3: Define your Radius Recipe configurations

> Radius provides a property `recipeConfig` which allows users to setup specific Git module sources for your Terraform Radius Recipes. Define this property in your Radius Environment resource, visit the [Radius Environment schema]({{< ref environment-schema >}}) page for more information.

Reviewer 3: Same thing here, specify if they need to add this to their `env.bicep` file

{{< rad file="snippets/env.bicep" embed=true marker="//ENV" >}}

> Note the keys listed inside of the `pat` property should match your path name to the secret store.

> 

Reviewer: Double blank lines here. Please make sure to only have one new line between sections.

## Step 4: Deploy your Radius Environment

Deploy your new Radius Environment:

```
rad deploy env.bicep
```

> 

Reviewer: How does the user "prove" that it worked? Maybe we can add steps here, or point users to a separate guide on how to use Recipes.

---
type: docs
title: "How-To: Pull Terraform modules from private git repositories"

> linkTitle: "Private git repos"

Reviewer: This is fine for now, but could you open a GitHub Issue to track splitting Terraform and Bicep Recipe documentation into two separate sections?

This will help avoid confusion where "Private git repos" might be associated with Bicep, which is unsupported.

Author: Yes already have that: https://github.com/radius-project/docs/issues/1077

description: "Learn how to setup your Radius environment to pull Terraform Recipe templates from a private git repository."
weight: 500
categories: "How-To"
tags: ["recipes", "terraform"]
---

This how-to guide will describe how to:

- Configure a Radius Environment to be able to pull Terraform Recipe templates from a private git repository.

### Prerequisites

Before you get started, you'll need to make sure you have the following tools and resources:

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
- [Radius Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- [Radius initialized with `rad init`]({{< ref howto-environment >}})

## Step 1: Create a personal access token

> Create a personal access token, this can be from [GitHub](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#about-personal-access-tokens), [GitLab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html), [Azure DevOps](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows) or any other Git platform.

Reviewer: ```suggestion
Create a personal access token, this can be from [GitHub](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#about-personal-access-tokens), [GitLab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html), [Azure DevOps](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows), or any other Git platform.
```

Oxford commas please

The PAT should have access to read the files inside the specific private repository.

## Step 2: Define a secret store resource

> Begin by configuring a [Radius Secret Store]({{< ref secretstore >}}) with the personal access token or username + password you previously created, which has access to your private git repository.

Reviewer: ```suggestion
Configure a [Radius Secret Store]({{< ref "/guides/author-apps/secrets/overview" >}}) with the personal access token or username + password you previously created, which has access to your private git repository.

> While this example shows a Radius-managed secret store where Radius creates the underlying secrets infrastructure, you can also bring your own existing secrets. Refer to the [secrets documentation]({{< ref "/guides/author-apps/secrets/overview" >}}) for more information.
```

Create a Bicep file `env.bicep`, import Radius, and  define your resource:

{{< rad file="snippets/env.bicep" embed=true marker="//SECRETSTORE" >}}

> > Note the property `pat` is a required key that refers to your personal access token, while `username` is an optional key you can specify if your git platform requires a username.

Reviewer: ```suggestion
> The property `pat` is required and refers to your personal access token or password, while `username` is optional and refers to a username, if your git platform requires one.
```

## Step 3: Configure Terraform Recipe git authentication

`recipeConfig` allows you to configure how Recipes should be setup and run. One available option is to specify git credentials for pulling Terraform Recipes from git sources. For more information refer to the [Radius Environment schema]({{< ref environment-schema >}}) page.

In your `env.bicep` file add an Environment resource, along with Recipe configuration which leverages the previously defined secret store for git authentication.

{{< rad file="snippets/env.bicep" embed=true marker="//ENV" >}}

## Step 4: Add a Terraform Recipe

Update your Environment with a Terraform Recipe, pointing to your private git repository. Note that your `templatePath` should contain a `git::` prefix, per the [Terraform module documentation](https://developer.hashicorp.com/terraform/language/modules/sources#generic-git-repository).

> {{< rad file="snippets/env-complete.bicep" embed=true marker="//ENV" markdownConfig="{linenos=table,hl_lines=[\"23-30\"],linenostart=30}" >}}

Reviewer: Could you double check this is highlighting correctly? On my Windows machine there needs to be one more semicolon highlighted, but I've seen line ending issues before on Unix/Windows so it might just be me.
<img width="593" alt="image" src="https://github.com/radius-project/docs/assets/54363786/5cc006e1-36e5-49eb-b4af-b0db59cca772">

## Step 5: Deploy your Radius Environment

Deploy your new Radius Environment:

```
rad deploy ./env.bicep -p pat=******
```

> ## Done

Reviewer 3: Recommend a different title for this section, perhaps something like "View the available Recipes in your newly deployed Radius Environment"

Reviewer: Jason and I discussed this and there's no way to test this due to a bug in the CLI. https://github.com/radius-project/radius/issues/7267

For now we'll just say done and when the above bug is fixed we'll update this document to verify the Recipe was added correctly.

> Your Radius Environment is now ready to utilize your Radius Recipes stored inside your private registry. For more information on Radius Recipes visit the [Recipes overview page]({{< ref "/guides/recipes/overview" >}}).

Reviewer 3: To @AaronCrawfis previous feedback, please add some steps that the user may follow to "prove" that their environment along with their private registry has been deployed.

Reviewer: Added a comment above, but due to a bug we can't. See https://github.com/radius-project/radius/issues/7267

{{< rad file="snippets/env.bicep" embed=true marker="//SECRETSTORE" >}}

> The property `pat` is required and refers to your personal access token or password, while `username` is optional and refers to a username, if your git platform requires one.

## Step 3: Configure Terraform Recipe git authentication

`recipeConfig` allows you to configure how Recipes should be setup and run. One available option is to specify git credentials for pulling Terraform Recipes from git sources. For more information refer to the [Radius Environment schema]({{< ref environment-schema >}}) page.

In your `env.bicep` file add an Environment resource, along with Recipe configuration which leverages the previously defined secret store for git authentication.

{{< rad file="snippets/env.bicep" embed=true marker="//ENV" >}}

## Step 4: Add a Terraform Recipe

Update your Environment with a Terraform Recipe, pointing to your private git repository. Note that your `templatePath` should contain a `git::` prefix, per the [Terraform module documentation](https://developer.hashicorp.com/terraform/language/modules/sources#generic-git-repository).

> {{< rad file="snippets/env-complete.bicep" embed=true marker="//ENV" markdownConfig="{linenos=table,hl_lines=[\"23-30\"],linenostart=30}" >}}

Author: ```suggestion
{{< rad file="snippets/env-complete.bicep" embed=true marker="//ENV" markdownConfig="{linenos=table,hl_lines=[\"23-31\"],linenostart=30}" >}}
```

> {{< rad file="snippets/env-complete.bicep" embed=true marker="//ENV" markdownConfig="{linenos=table,hl_lines=[\"23-31\"],linenostart=30}" >}}

Reviewer: This isn't rendering correctly on the staging site: 
![image](https://github.com/radius-project/docs/assets/54363786/7185694a-5066-45ec-86c0-075959065d51)
https://calm-bush-0cc2a4f1e-1073.westus2.3.azurestaticapps.net/guides/recipes/terraform/howto-private-registry/#step-4-add-a-terraform-recipe

Author: Fixed: 
![Screenshot 2024-03-05 at 1 33 29 PM](https://github.com/radius-project/docs/assets/83607984/9eb325d5-f8cf-4549-87ed-5d287691cb12)

## Step 5: Deploy your Radius Environment

Deploy your new Radius Environment:

```

> rad deploy ./env.bicep -p pat=******

Reviewer: Did you test this?

Author: I tested out the deploy command but not with the pat being passed, however when I synced with @vishwahiremat  I could not get my local or codespace to work correctly but he confirmed that the deploy command works on his machine.

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
- [Radius Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

> - A [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#about-personal-access-tokens) for your Git repository.

Reviewer: This shouldn't be specific to just GitHub, as Azure DevOps, GitLab, and other git platforms are also supported. I'd make this Step 1, with links to all the supported platforms.

## Step 1: Define a secret store resource

Begin by configuring a [Radius Secret Store]({{< ref secretstore >}}) with the personal access token or username + password you previously created, which has access to your private git repository.

> Create a Bicep file `env.bicep`, import Radius and  define your resource:

Reviewer: ```suggestion
Create a Bicep file `env.bicep`, import Radius, and  define your resource:
```

For consistency let's use Oxford commas in comma-separated lists.

{{< rad file="snippets/env.bicep" embed=true marker="//SECRETSTORE" >}}

> Note the property `pat` is a required property that refers to your personal access token, while `username` can be optional.

> ## Step 2: Define your Radius Recipe configurations

Reviewer: ```suggestion
## Step 2: Configure Terraform Recipe git authentication
```

> Radius provides a property `recipeConfig` which allows users to setup specific Git module sources for your Terraform Radius Recipes. In your `env.bicep` file define this property in your Radius Environment resource, visit the [Radius Environment schema]({{< ref environment-schema >}}) page for more information.

Reviewer: ```suggestion
`recipeConfig` allows you to configure how Recipes should be setup and run. One available option is to specify git credentials for pulling Terraform Recipes from git sources. For more information refer to the [Radius Environment schema]({{< ref environment-schema >}}) page.

In your `env.bicep` file add an Environment resource, along with Recipe configuration which leverages the previously defined secret store for git authentication.
```

{{< rad file="snippets/env.bicep" embed=true marker="//ENV" >}}

> > Note the keys listed inside of the `pat` property should match your path name to the secret store.

Reviewer: I'm not following what this means.
